File: Hw3/sol1.py
class Solution():
    def LCSHelper(self,A,B):
        n=len(A)
        m=len(B)
        memo=[[0]*(m+1) for _ in range(n+1)]
        for i in range(n+1):
            for j in range(m+1):
                if (i==0) or (j==0):
                    memo[i][j]=0
                elif A[i-1]==B[j-1]:
                    memo[i][j]=memo[i-1][j-1]+1
                else:
                    memo[i][j]=max(memo[i-1][j],memo[i][j-1])
        return memo
    def longestCommonSubsequence(self,A,B):
        self.memo= self.LCSHelper(A,B)
        print ("Length of LCS is ",self.memo[-1][-1])
        common_subseq=self.reconstruct(A,B)
        print ("LCS is ",common_subseq)
        return self.memo
    def reconstruct(self,A,B):
        n=len(A)
        m=len(B)
        row=n
        col=m
        LCS=""
        while(row>0 and col>0):
            if A[row-1] == B[col-1]:
                LCS=A[row-1]+LCS
                row=row-1
                col=col-1
            elif self.memo[row-1][col] > self.memo[row][col-1]:
                row=row-1
            else:
                col=col-1
        return LCS


sol_obj = Solution()

C = "abcd"
D = "acd"

sol_obj.longestCommonSubsequence(C, D)

File: Hw3/sol1.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <vector>
#include <list>
#include <set>
#include <chrono>
#include <numeric>
using namespace std;
using namespace std::chrono;

void print_2D_vector(vector<vector<int>> const &v) {
  //helper function to print 2D vector
for (vector<int> row: v) {
    for (int val: row) {
      cout << val << " ";
}
    cout << '\n';
  }
}

vector<vector<int>> LCSHelper(string A,string B){
  int n=A.size();
  int m=B.size();
  vector<vector<int>> memo(n+1, vector<int>(m+1, 0));
  for (int i=0;i<n+1;i++){
    for (int j=0;j<m+1;j++){
      if ((i==0) or (j==0)){
        memo[i][j]=0;
      }
      else if (A[i-1]==B[j-1]){
        memo[i][j]=memo[i-1][j-1]+1;
      }
      else{
        memo[i][j]=max(memo[i-1][j],memo[i][j-1]);
      }
    }
  }
  return memo;
}
string reconstruct(string A,string B,vector<vector<int>> memo){
  int n=A.size();
  int m=B.size();
  int row=n;
  int col=m;
  string LCS="";
  while(row>0 and col>0){
    if (A[row-1] == B[col-1]){
      LCS=A[row-1]+LCS;
      row=row-1;
      col=col-1;
    }else if (memo[row-1][col] > memo[row][col-1]){
      row=row-1;
    }else{
      col=col-1;
    }
  }
  return LCS;
}
void longestCommonSubsequence(string A,string B){
  vector<vector<int>> memo=LCSHelper(A,B);
  cout << "Length of LCS is : "<<memo[A.size()][B.size()]<<endl;
  string common_subseq=reconstruct(A,B,memo);
  cout << "LCS is : "<<common_subseq<<endl;
}

int main(int argc, char const *argv[]) {
  string C = "ABCDGH";
  string D = "AEDFHR";
  longestCommonSubsequence(C,D);
  return 0;
}
